WIP: Change integer constants to enum in Conditional

danielb987 opened the pull request.

bobjacobsen commented:
Travis graphical failed with messages:
```
[ERROR] Errors: 
[ERROR]   LogixTableActionTest.testTreeEditor:112 » MissingResource Resource 'ItemTypeNo...
[ERROR]   ConditionalListEditTest.addConditionalTest:41 » TimeoutExpired Frame with titl...
[ERROR]   ConditionalTreeEditTest.addConditionalTest:32 » MissingResource Resource 'Item...
[ERROR]   ConditionalTreeEditTest.comboBoxTest:117 » MissingResource Resource 'ItemTypeN...
[ERROR]   ConditionalTreeEditTest.singlePickTest:84 » MissingResource Resource 'ItemType...
```
The timeout might be a false positive, but the missing resources are probably real.

pabender commented:
> On Oct 3, 2018, at 6:48 AM, Bob Jacobsen <notifications@github.com> wrote:
> Travis graphical failed with messages:
> 
> [ERROR] Errors: 
> [ERROR]   LogixTableActionTest.testTreeEditor:112 » MissingResource Resource 'ItemTypeNo...
> [ERROR]   ConditionalListEditTest.addConditionalTest:41 » TimeoutExpired Frame with titl...
> [ERROR]   ConditionalTreeEditTest.addConditionalTest:32 » MissingResource Resource 'Item...
> [ERROR]   ConditionalTreeEditTest.comboBoxTest:117 » MissingResource Resource 'ItemTypeN...
> [ERROR]   ConditionalTreeEditTest.singlePickTest:84 » MissingResource Resource 'ItemType...
> The timeout might be a false positive, but the missing resources are probably real.
> 
I have not looked at the actual log, but the timeout there appears to be coming from Jemmy, and is saying a frame with a specified title can’t be found.  This error will occur if the window either never appears ( it must be visible ) or if the title doesn’t match what was specified in the Jemmy operator constructor call.

pabender commented:
Appveyor failed with what appears to be the same errors.  It shows the full text of the timeout as:

    [junit] Testcase: addConditionalTest(jmri.jmrit.conditional.ConditionalListEditTest):	Caused an ERROR
    [junit] Frame with title "Edit Variable" (WindowWaiter.WaitWindowTimeout)
    [junit] org.netbeans.jemmy.TimeoutExpiredException: Frame with title "Edit Variable" (WindowWaiter.WaitWindowTimeout)
    [junit] 	at org.netbeans.jemmy.Waiter.waitAction(Waiter.java:213)
    [junit] 	at org.netbeans.jemmy.WindowWaiter.waitWindow(WindowWaiter.java:437)
    [junit] 	at org.netbeans.jemmy.WindowWaiter.waitWindow(WindowWaiter.java:227)
    [junit] 	at org.netbeans.jemmy.FrameWaiter.waitFrame(FrameWaiter.java:243)
    [junit] 	at org.netbeans.jemmy.operators.FrameOperator.waitFrame(FrameOperator.java:431)
    [junit] 	at org.netbeans.jemmy.operators.JFrameOperator.<init>(JFrameOperator.java:90)
    [junit] 	at org.netbeans.jemmy.operators.JFrameOperator.<init>(JFrameOperator.java:126)
    [junit] 	at org.netbeans.jemmy.operators.JFrameOperator.<init>(JFrameOperator.java:143)
    [junit] 	at org.netbeans.jemmy.operators.JFrameOperator.<init>(JFrameOperator.java:158)
    [junit] 	at jmri.jmrit.conditional.ConditionalListEditTest.addConditionalTest(ConditionalListEditTest.java:41)

danielb987 commented:
@bobjacobsen @pabender 

Thanks! I forgot to add some bundles and adding them solved the problem.

danielb987 commented:
@bobjacobsen @pabender 

What is the difference between `SIGNAL_HEAD_RED` to `SIGNAL_HEAD_FLASHGREEN` on one hand, and `SIGNAL_HEAD_LIT`, `SIGNAL_HEAD_HELD` and `SIGNAL_HEAD_APPEARANCE_EQUALS` on the other hand?

`ConditionalListEdit` and `ConditionalTreeEdit` has this code segment and I think it's fragile, so I would like to maybe add a method `is???()` in `Conditional.Type`, but before that, I need to understand the code.
```
if ((Conditional.Type.SIGNAL_HEAD_RED.getIntValue() <= testType.getIntValue()
        && testType.getIntValue() <= Conditional.Type.SIGNAL_HEAD_FLASHGREEN.getIntValue())
        || Conditional.Type.SIGNAL_HEAD_LUNAR == testType
        || Conditional.Type.SIGNAL_HEAD_FLASHLUNAR == testType) {
```
The reason I think it's fragile is that the code depends on the integer values of the constants.

danielb987 commented:
I have found a bug in `RouteTableAction` at line 1539.

I changed `ConditionalVariable(not, opern, type, name, trigger)` to use the enum `Conditional.Type` instead of an `int` for the paramater `type`.

After that, I was able been able to do the change everythere, except in one place in `RouteTableAction` at line 1539. The code was:
```
int type = Turnout.CLOSED;
if (mode == Route.ONTHROWN) {
    type = Turnout.THROWN;
}
variableList.add(new ConditionalVariable(false, Conditional.Operator.NONE,
        type, turnoutLockSystemName, true));
```
These constants have the following values:
```
Turnout.CLOSED = 2
Turnout.THROWN = 4
```
But the constants in `Conditional` have the values:
```
Conditional.TYPE_TURNOUT_THROWN = 3
Conditional.TYPE_TURNOUT_CLOSED = 4
```
I have changed the code to:
```
Conditional.Type conditionalType = Conditional.Type.TURNOUT_CLOSED;
if (mode == Route.ONTHROWN) {
    conditionalType = Conditional.Type.TURNOUT_THROWN;
}
variableList.add(new ConditionalVariable(false, Conditional.Operator.NONE,
        conditionalType, turnoutLockSystemName, true));
```
I changed the name of the variable `type` since it is used in more places in that method as an `int`.

dsand47 commented:
A little background.

The 36 Types map to an object type **_and_** state.  So type 3 says ``if turnout X is thrown``.  This translates to ``if X.getKnownState() == Turnout.THROWN``.

The 53 Actions map to an object type and the defined action.  For example, 10 is ``delayed set sensor`` while 9 is just ``set sensor``.

Reading Conditional XML is challenging since all you get are numbers.

The SIGNAL HEAD example posted earlier says ``if testType == 11 to 17 or 21 or 22 then  ...``.  It would be nice to be able to say ``if getSignalHeadAppearanceTypes().contains(testType)``.

danielb987 commented:
I did some work on a branch based on this branch, but when I tried to merge that branch into this branch something got wrong. So I close this PR and open a new PR based on the other branch. That branch contains all the good changes from this branch and the new changes.
